### PR TITLE
Use ImageMagick to generate LVGL image source code

### PR DESCRIPTION
**Pull request by eendeego**
Instead of using the Web UI to upload image / download source code, add an `art` directory and configuration to:
  * Manage a list of files and their conversion to 68x140 1bpp preferences
  * Generate the LVGL source code
  * Integrate all of this in the build

### CODE REVIEW

File: build.sh [hunk 1]
```diff
@@ -2,6 +2,27 @@
 
 set -euo pipefail
 
+header() {
+    local args top_line bottom_line
+    args=(${@})
```
Comment (soarez) [inline build.sh]: This is unused

File: build.sh [hunk 2]
```diff
@@ -2,6 +2,27 @@
...
+    if (( $top_line )); then
+        echo -e "\n========================================================================"
+    fi
+
+    echo == "${@}"
+
+    if (( $bottom_line )); then
+        echo -e "========================================================================\n"
+    fi
+}
```
Comment (soarez) [inline build.sh]:
```suggestion
header() {
    local bar='========================================================================'
    local top="\n${bar}\n"
    local bottom="\n${bar}\n"
    local middle
    while [[ "$1" == -* ]]; do
        case $1 in
            --top|-T) top=''; ;;
            --bottom|-B) bottom=''; ;;
            *) >&2 echo "ERROR: Unknown flag $1"; exit 0; ;;
        esac
        shift
    done
    middle="== ${*}"
    if [[ -n "$top" ]]; then
        middle="${middle} ${bar}"
        middle="${middle:0:${#bar}}"
    fi
    echo -e "${top}${middle}${bottom}"
}
```

File: build.sh [hunk 3]
```diff
@@ -18,12 +41,28 @@ zephyr_sdk_path="${this_dir}/zephyr-sdk-${zephyr_version}/"
...
 if [[ ! -d "$zephyr_sdk_path" ]]; then
+    header -B "Installing zephyr-sdk ==============================================="
+
     if [[ ! -f "$zephyr_archive" ]]; then
+        header -T -B "  Fetching zephyr-sdk"
         curl -LO "https://github.com/zephyrproject-rtos/sdk-ng/releases/download/v${zephyr_version}/$zephyr_archive"
     fi
-    tar xf "$zephyr_archive"
+
+    header -T -B "  Untarring zephyr-sdk"
```
Comment (soarez) [inline build.sh:52]: Had never seen this term before - untarring

File: build.sh [hunk 4]
```diff
@@ -90,15 +125,17 @@ if [ -e zephyr/module.yml ]; then
     >&2 echo "ERROR: Branch not implemented!"; exit 1
 fi
 
+header "Building All the things!"
+
+build_init
+
+sh ./patch_art.sh
```
Comment (soarez) [inline build.sh:133]: Should this be part of `build_init`?

File: patch_art.sh [hunk 5]
```diff
@@ -0,0 +1,146 @@
+#!/bin/bash
+
+ART_FILE=zmk/app/boards/shields/nice_view/widgets/art.c
+PERIPHERAL_STATUS_FILE=zmk/app/boards/shields/nice_view/widgets/peripheral_status.c
```
Comment (soarez) [inline patch_art.sh]:
```suggestion
this_dir="$(cd "$(dirname "$0")" && pwd)"
widgets_dir="${this_dir}/zmk/app/boards/shields/nice_view"
ART_FILE="${widgets_dir}/art.c"
PERIPHERAL_STATUS_FILE="${widgets_dir}/peripheral_status.c
```

File: patch_art.sh [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+reset_zmk() {
+    cd zmk
```
Comment (soarez) [inline patch_art.sh]:
```suggestion
    cd "${this_dir}/zmk"
```

File: patch_art.sh [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+    cd ..
+}
+
+generate_lvgl_img() {
+    local name all_caps
+    name=$1
+    src=$2
+
+    all_caps=$(echo "$name" | tr "[:lower:]" "[:upper:]")
```
Comment (soarez) [inline patch_art.sh]:
```suggestion
    all_caps=$(tr "[:lower:]" "[:upper:]" <<<"$name")
```

Another option is:
```
all_caps="${name^^}"
```

File: patch_art.sh [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    git restore .
+    git clean -fd
+
+    cd ..
+}
+
+generate_lvgl_img() {
+    local name all_caps
+    name=$1
+    src=$2
```
Comment (soarez) [inline patch_art.sh]:
```suggestion
    local name="$1"
    local src="$2"
```

See https://www.shellcheck.net/wiki/SC2155 for why declaration and assignment is sometimes separate.

File: patch_art.sh [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+magick_args() {
+    local name tmp_dir preview file resize color_convert magick_args
+    name=$1
+    tmp_dir=$2
+    preview=$3
+
+    read -d "\t" -r file resize color_convert magick_args \
+        <<< $(jq -r ".images[] | select(.name == \"${name}\") | [(.file // false), (.resize // false), (.color_convert // \"none\"), (.magick_args // \"-\")] | @tsv" art/config.json)
```
Comment (soarez) [inline patch_art.sh:63]: nice way to convert json to shell variables

File: patch_art.sh [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        args+="$magick_args"
+    fi
+
+    if [[ "$preview" != "-preview" ]]; then
+        args+=("-rotate" "90" "$tmp_dir/$name.pbm")
+    else 
+        args+=("art/preview/${name}.png")
+    fi
+
+    echo ${args[@]}
```
Comment (soarez) [inline patch_art.sh]: Need to use quotes to deal with possible spaces in file paths
```suggestion
    echo "${args[@]}"
```

**Review by soarez — COMMENTED** [review]
Thanks for the PR!

File: build.sh (hunk 3 above)
Comment (eendeego) [reply build.sh:52]: Google says it's very common in this context.

File: build.sh (hunk 4 above)
Comment (eendeego) [reply build.sh:133]: My understanding is that `build_init` sets up the build, e.g. directories and all, and that `patch_art` is already one of the steps of the build.

File: build.sh (hunk 2 above)
Comment (eendeego) [reply build.sh]: I was thinking on how to add the = signs on the middle line, but my script-fu didn't take me there quickly.

**Comment by soarez** [comment]
Thanks @eendeego !

**Pull request merged**